src/mcsat/solver.c

```diff
@@ -2339,32 +2342,88 @@ bool mcsat_decide_assumption(mcsat_solver_t* mcsat, model_t* mdl, uint32_t n_ass
+  assert(var != variable_null);
+  assert(!trail_has_value(mcsat->trail, var));
 
   uint32_t i;
-  variable_t var;
   plugin_t* plugin;
   plugin_trail_token_t decision_token;
+  bool made_decision = false;
+
+  // Get the owner that will decide that value of the variable
+  i = mcsat->decision_makers[variable_db_get_type_kind(mcsat->var_db, var)];
+  assert(i != MCSAT_MAX_PLUGINS);
+  // Construct the token
+  trail_token_construct(&decision_token, mcsat->plugins[i].plugin_ctx, var);
+  // Decide
+  if (trace_enabled(mcsat->ctx->trace, "mcsat::decide")) {
+    mcsat_trace_printf(mcsat->ctx->trace, "mcsat_decide(): with %s\n", mcsat->plugins[i].plugin_name);
+    mcsat_trace_printf(mcsat->ctx->trace, "mcsat_decide(): variable ");
+    variable_db_print_variable(mcsat->var_db, var, trace_out(mcsat->ctx->trace));
+    mcsat_trace_printf(mcsat->ctx->trace, "\n");
+  }
+  plugin = mcsat->plugins[i].plugin;
+
+  // Ask the owner to decide
+  mcsat_push_internal(mcsat);
+  assert(plugin->decide);
+  plugin->decide(plugin, var, (trail_token_t*) &decision_token, force_decision);
 
-  ivector_t skipped_variables; // Variables we took from the queue but plugin didn't decide
-  init_ivector(&skipped_variables, 0);
+  if (decision_token.used == 0) {
+    // If not decided, remember and go on
+    made_decision = false;
+    mcsat_pop_internal(mcsat);
+  } else {
+    made_decision = true;
+    // Decided, we can continue with the search
+    (*mcsat->solver_stats.decisions)++;
+    // Plugins are not allowed to cheat and decide on another variable
+    assert(trail_has_value(mcsat->trail, var));
+    if (trace_enabled(mcsat->ctx->trace, "mcsat::decide")) {
+      FILE* out = trace_out(mcsat->ctx->trace);
+      fprintf(out, "mcsat_decide(): value = ");
+      mcsat_value_print(trail_get_value(mcsat->trail, var), out);
+      fprintf(out, "\n");
+    }
+  }
+
+  // a forced decision implies a made decision
+  assert(!force_decision || made_decision);
+  return made_decision;
+}
+
+/**
+ * Decides a variable using one of the plugins. Returns true if a variable
+ * has been decided, or a conflict detected.
+ */
+static
+bool mcsat_decide(mcsat_solver_t* mcsat) {
 
   assert(!mcsat->trail->inconsistent);
 
+  ivector_t skipped_variables; // Variables we took from the queue but plugin didn't decide
+  init_ivector(&skipped_variables, 0);
+
+  variable_t var;
+  bool aux_choice; // indicates that var was not taken from the queue
   bool force_decision = false;
+  double rand_freq = mcsat->heuristic_params.random_decision_freq;
+
   while (true) {
     var = variable_null;
+    aux_choice = true;
 
     // Us the top variables first
-    for (i = 0; i < mcsat->top_decision_vars.size; ++i) {
+    for (uint32_t i = 0; i < mcsat->top_decision_vars.size; ++i) {
       var = mcsat->top_decision_vars.data[i];
       assert(var != variable_null);
       if (!trail_has_value(mcsat->trail, var)) {
-        force_decision = true;
```

Comment:
why remove this?

Reply:
Because it was only a quickfix to avoid getting skipped variables to be re-inserted in the var queue.
Not forcing is the preferred option, as it gives the plugins more flexibility.

Reply:
I think these variables need to be forced because they are top vars or in the var order given by the user.